### PR TITLE
Added cuDNN v5 support & Added optional dependencies to setup.py

### PR DESCRIPTION
**Pull request by mgermain**
and some bug fixes

### CODE REVIEW

**Comment by nouiz** [comment]
travis is failing.

**Comment by nouiz** [comment]
travis error is:

/home/travis/build/Theano/Theano/theano/sandbox/cuda/dnn.py:docstring of theano.sandbox.cuda.dnn.GpuDnnConv3d:11: WARNING: Definition list ends without a blank line; unexpected unindent.

**Comment by lamblin** [comment]
@mgermain Are there still any issues when running the tests on a machine with a GPU?

**Comment by mgermain** [comment]
Not in the test_dnn and test_abstractconv but I'll run the whole thing with
both cuDNN 4 and 5 and report here.

**Comment by mgermain** [comment]
All good.

File: theano/sandbox/cuda/dnn.py [hunk 1]
```diff
@@ -384,8 +387,11 @@ def get_op_params(self):
                 # need v3
                 alg = 'CUDNN_CONVOLUTION_FWD_ALGO_FFT'
             elif self.algo == 'fft_tiling':
-                # need v4
+                # need v4 for conv2d, need v5 for conv3d
                 alg = 'CUDNN_CONVOLUTION_FWD_ALGO_FFT_TILING'
+            elif self.algo == 'winograd':
+                # need v4 for conv2d, need v5 for conv3d
```
Comment (nouiz) [inline theano/sandbox/cuda/dnn.py]: Update this comment. Always need v5.

**Pull request merged**